### DISCUSSION TITLE
How can I ignore warning msg in vips_save_buffer

### DISCUSSION THREAD

**Discussion by jishi92** [libvips/libvips#3222] [Q&A] [answered]
In this case, image can be process successfully.
But an exception error is thrown：
```
VipsJpeg: Premature end of input file
```

How can I ignore warning msg in vips_jpegsave_buffer in my code.I think the result is valid and I don't want to accept this warning.

The cmd is : 
```
vips invert broken-jpg.jpeg test000.jpg
```
![broken-jpg](https://user-images.githubusercontent.com/22196414/206409314-249ba272-d945-4f43-a415-60b1f6a9fe9b.jpeg)

**Comment by jcupitt** [comment, marked as answer]
Hi again,

Warnings have no effect by default. For example:

```
$ vips copy truncated.jpg x.jpg

(vips:124623): VIPS-WARNING **: 11:13:08.580: read gave 2 warnings

(vips:124623): VIPS-WARNING **: 11:13:08.581: VipsJpeg: Premature end of JPEG file

$ echo $?
0
```

So the command succeeded, despite the warning.

You can make it fail with `fail-on`, eg.:

```
 $ vips copy truncated.jpg[fail-on=warning] x.jpg

(vips:124761): VIPS-WARNING **: 11:14:39.724: error in tile 0 x 136
VipsJpeg: Premature end of input file
$ echo $?
1
```

**Reply by jishi92** [reply]
Yes,I know this param and I have make fail_on with:
```
fail_on=VIPS_FAIL_ON_ERROR
```
But, I still get the warning msg.

```
[root@3a68b95ade7b Downloads]# vips copy broken-jpg.jpeg[fail-on=error] xxx.jpeg

(vips:13392): VIPS-WARNING **: 20:23:23.975: error in tile 0 x 160

(vips:13392): VIPS-WARNING **: 20:23:23.975: error in tile 0 x 168
```

I think the fail_on param is affect on load_image,but the warning msg was thrown in save_buffer.
I don't know if I'm getting that right

**Reply by jishi92** [reply]
```
[root@3a68b95ade7b Downloads]# vips copy broken-jpg.jpeg[fail-on=none] xxx.jpeg

(vips:13405): VIPS-WARNING **: 20:28:41.145: read gave 2 warnings

(vips:13405): VIPS-WARNING **: 20:28:41.145: VipsJpeg: Premature end of JPEG file
```

**Reply by jcupitt** [reply]
They are just warnings, you can ignore them. The operation has completed successfully.

**Reply by jishi92** [reply]
Yes,my question is how to ignore them in my code.
Because save_buffer return 1 not 0 in libvips.
I don't know how to do it.

**Reply by jcupitt** [reply]
It should return 0. I made you a test program:

```C
// gcc -g -Wall try337.c `pkg-config vips --cflags --libs`

#include <vips/vips.h>

int 
main(int argc, char *argv[])
{
        void *buf, *out_buf;
        size_t len, out_len;
        VipsImage *image;

        if (VIPS_INIT(argv[0]))
                vips_error_exit(NULL);

        if (!g_file_get_contents(argv[1], (void *) &buf, &len, NULL))
                vips_error_exit("unable to load");

        printf ("loaded %zd bytes data\n", len);
        
        if (!(image = vips_image_new_from_buffer(buf, len, "", NULL)))
                vips_error_exit(NULL);
        
        if (vips_image_write_to_buffer(image, 
                ".jpg", &out_buf, &out_len, NULL))
                vips_error_exit(NULL);

        printf ("saved as %zd bytes of jpeg\n", out_len);
        
        g_object_unref(image);
        free(buf);
        free(out_buf);

        return 0;
}
```

I can make a test image like this:

```
$ head -c 20000 ~/pics/k2.jpg > x.jpg
```

So take just the first 20,000 bytes of a jpg image, then:

```
$ ./a.out x.jpg
loaded 20000 bytes data
saved as 63299 bytes of jpeg
```

You need to post a sample image and program that shows your error, or I will not be able to reproduce the problem.

**Reply by jishi92** [reply]
Hi,Thanks for your quickly reply.
I use govips and I find this function return 1.
```
int save_buffer(const char *operationName, SaveParams *params,
                SetSaveOptionsFn setSaveOptions) {
  VipsBlob *blob;
  VipsOperation *operation = vips_operation_new(operationName);
  if (!operation) {
    return 1;
  }

  if (vips_object_set(VIPS_OBJECT(operation), "in", params->inputImage, NULL)) {
    return 1;
  }

  if (setSaveOptions(operation, params)) {
    g_object_unref(operation);
    return 1;
  }

  if (vips_cache_operation_buildp(&operation)) {       
    vips_object_unref_outputs(VIPS_OBJECT(operation));
    g_object_unref(operation);
    return 1;           //here return 1.
  }

  g_object_get(VIPS_OBJECT(operation), "buffer", &blob, NULL);
  g_object_unref(operation);

  VipsArea *area = VIPS_AREA(blob);

  params->outputBuffer = (char *)(area->data);
  params->outputLen = area->length;
  area->free_fn = NULL;
  vips_area_unref(area);

  return 0;
}
```
https://github.com/davidbyttow/govips/blob/master/vips/foreign.c#L200


I'm not very familiar with C language, so I don't know how to ignore it.
Could you help me see it.

**Reply by jcupitt** [reply]
Sorry, I don't understand what's going wrong, and I can't reproduce the error.

Maybe ask on the govips repository?

**Reply by jishi92** [reply]
Ok,I will try do it though the maintainers of govips are not very active.
And I'm going to continue to work on that.

**Reply by jishi92** [reply]
Hi @jcupitt :

In my tests, I found that it only return 0  if I set fail_on=none.
Vips will return 1  doesn't matter if I set fail_on=error or fail_on=truncated.

```
[root@3a68b95ade7b Downloads]# vips copy broken-jpg.jpeg[fail-on=error] xxxx11.jpg

(vips:32566): VIPS-WARNING **: 17:35:59.555: error in tile 0 x 160

(vips:32566): VIPS-WARNING **: 17:35:59.555: error in tile 0 x 168
VipsJpeg: Premature end of input file
[root@3a68b95ade7b Downloads]# echo $?
1
[root@3a68b95ade7b Downloads]# vips copy broken-jpg.jpeg[fail-on=none] xxxx22.jpg

(vips:36017): VIPS-WARNING **: 17:49:37.154: read gave 2 warnings

(vips:36017): VIPS-WARNING **: 17:49:37.154: VipsJpeg: Premature end of JPEG file

[root@3a68b95ade7b Downloads]# echo $?
0

```



I don't understand this result. Can you help me explain it.

**Reply by jcupitt** [reply]
Could you upload this image somewhere? You can't use github, unfortunately, they will alter it.

Also, which libvips version are you using?

**Reply by jishi92** [reply]
My env:
centos8.4 and vips-8.13.0.
I can email you later.

**Reply by jcupitt** [reply]
I tried with your image:

```
john@kiwi:~/pics$ vips --version
vips-8.13.0
john@kiwi:~/pics$ vips copy broken-jpg.jpeg x.jpg

(vips:27002): VIPS-WARNING **: 16:43:51.797: read gave 2 warnings

(vips:27002): VIPS-WARNING **: 16:43:51.797: VipsJpeg: Premature end of JPEG file

john@kiwi:~/pics$ echo $?
0
```

And it's all working for me. Are you certain you are running 8.13.0?

**Reply by jishi92** [reply]
I'm sure.
I test again and find because of [fail_on] params.
I add it and you didn't.

**Reply by jishi92** [reply]
I guess I found the reason. It is a bug caused by fail_on parameter.

**Reply by jcupitt** [reply]
I don't think there's a bug. Your jpg file is truncated and also triggers an error. 

If you set `fail-on=none` (this is the default), it returns successfully and prints some warnings. If you set `fail-on=warning` or `truncated`, it will fail with an error message.

**Reply by jishi92** [reply]
Thank you jcupitt.
That's true, the default fail-on=none.I misread it earlier and thought the default value was warning.
I don't understand why vips still return 1  if I set fail-on=error, those msg look like just warning.
Are they errors instead of warnings？

By the way, 
If I set fail-on=none, will vips no longer return any errors?
I don't think it's very friendly for a image processor.
How can I optimize my program？
Or what value should I set for this parameter?

**Reply by jcupitt** [reply]
libjpeg can flag an error with a truncated file -- it depends exactly where the truncation happens. And it's not always possible to detect a truncated file.

The `fail-on` parameter just affects the behaviour of the loader. You'll still get errors from other operations.

**Reply by jishi92** [reply]
Yes, I think so.The fail-on parameter just affects the behaviour of the loader.
So why does this parameter(fail-on=none) affect the exported result.
This parameter should just affect load_image not save_image.

**Reply by jcupitt** [reply]
It does just affect the load. But because libvips is demand-driven, the load won't actually occur until the save operation executes, so you see the error later.

**Reply by jishi92** [reply]
Ok,I think I get it.
If I set fail-on=none does that mean vips will not return any errors(always return 0), whether or not the processing succeeds.

**Reply by jcupitt** [reply]
You'll still get errors, just not load errors.

**Reply by jishi92** [reply]
That means vips loads any images successfully and it's still possible to throw an error in save_image or any other operations.

**Comment by jishi92** [comment]


**Reply by jcupitt** [reply]
Could you open a new discussion for a new question please?

It will make it easier for other users to find answers.

**Reply by jishi92** [reply]
ok ,thanks
https://github.com/libvips/libvips/discussions/3260
